File: frontend/src/features/announcements/api/announcement.service.ts
import apiClient from "@/lib/api";
import { PageResponse } from "@/types/common";
import {
  Announcement,
  CreateAnnouncementRequest,
} from "@/features/announcements/types";

export const announcementService = {
  /**
   * Lấy danh sách Thông báo.
   * Hỗ trợ nhiều bộ lọc: từ khóa, năm học, khoa, loại thông báo.
   */
  getAll: async (
    keyword?: string,
    page = 0,
    size = 10,
    academicYearId?: number,
    departmentId?: number,
    isSystem?: boolean,
  ): Promise<PageResponse<Announcement>> => {
    const params = new URLSearchParams();
    if (keyword) params.append("keyword", keyword);
    if (academicYearId)
      params.append("academicYearId", academicYearId.toString());
    if (departmentId) params.append("departmentId", departmentId.toString());
    if (isSystem !== undefined) params.append("isSystem", isSystem.toString());
    params.append("page", page.toString());
    params.append("size", size.toString());

    const response = await apiClient.get<PageResponse<Announcement>>(
      `/announcements?${params.toString()}`,
    );
    return response.data;
  },

  /**
   * Tạo Thông báo mới.
   */
  create: async (data: CreateAnnouncementRequest): Promise<Announcement> => {
    const response = await apiClient.post<Announcement>("/announcements", data);
    return response.data;
  },

  /**
   * Cập nhật Thông báo.
   */
  update: async (
    id: number,
    data: CreateAnnouncementRequest,
  ): Promise<Announcement> => {
    const response = await apiClient.put<Announcement>(
      `/announcements/${id}`,
      data,
    );
    return response.data;
  },

  /**
   * Xóa Thông báo.
   */
  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/announcements/${id}`);
  },
};

File: backend/src/main/java/vn/edu/husc/researchhub/service/AnnouncementService.java
package vn.edu.husc.researchhub.service;

import org.springframework.data.domain.Page;
import vn.edu.husc.researchhub.dto.request.AnnouncementRequest;
import vn.edu.husc.researchhub.dto.response.AnnouncementResponse;

/**
 * Service quản lý Thông báo.
 */
public interface AnnouncementService {
  /**
   * Lấy danh sách thông báo có phân trang.
   * Hỗ trợ lọc theo nhiều tiêu chí: từ khóa, năm học, khoa, loại thông báo.
   */
  Page<AnnouncementResponse> getAll(
      String keyword,
      int page,
      int size,
      Integer academicYearId,
      Integer departmentId,
      Boolean isSystem);

  /**
   * Tạo thông báo mới.
   */
  AnnouncementResponse create(AnnouncementRequest request);

  /**
   * Cập nhật thông báo.
   */
  AnnouncementResponse update(Integer id, AnnouncementRequest request);

  /**
   * Xóa thông báo.
   */
  void delete(Integer id);
}
